src/box/lua/net_box.lua

```diff
@@ -1078,6 +1197,40 @@ local function check_eval_args(args)
     end
 end
 
+function remote_methods:stream(stream_id)
+    check_remote_arg(self, 'stream')
+    if type(stream_id) ~= 'number' and type(stream_id) ~= 'nil' then
+        box.error(E_PROC_LUA, "Use remote:stream(stream_id) with an " ..
+                  "integer stream_id or without any arguments")
+    end
+    local last_sid = self._last_stream_id
+    if (stream_id and last_sid > 0) or (not stream_id and last_sid < 0) then
+        box.error(E_PROC_LUA, "You can use stream_id, which is generated " ..
+                  "automatically, or set it manually, but do not mix these " ..
+                  "two options")
+    end
+    self._last_stream_id = stream_id and -1 or (last_sid + 1)
+    stream_id = stream_id and stream_id or self._last_stream_id
+
+    if self._streams[stream_id] then
+        return self._streams[stream_id]
+    end
+
+    local stream = setmetatable({
+        _stream_id = stream_id,
+        space = setmetatable({
+            _stream = nil,
+        }, stream_spaces_mt),
+        _conn = self,
+        _weak = setmetatable({stream = nil}, weak_value_mt),
+        _transport = self._transport,
+    }, stream_mt )
```

Comment:
remote_methods._request also accesses self.state, which is nil in case of stream, so that piece of logic is broken, no?

Reply:
Yes it's my mistake. But in general, there is a question of how best to implement this:
First, for each thread, i set the value  _ index = = connection, but later i decide to
implement stream_methods which contains only necessary methods from connection. 
This implementation has disadvantages that you have described below, but earlier we needed to make a bunch of dummy methods, for those methods that the stream should not have (for example 'stream' method from connection.)

Reply:
Discussed f2f. Agreed to try to make stream a child of connection (set __index = connection). Child stream will have to set _stream_id (for connection it will be nil or 0) and override new_stream() method. Dummy methods are not needed - we want stream to have exactly the same API as connection.

Advantages:
 - Better test coverage, because tests written for net.box connection will test all the code executed by streams except setting stream_id, which should be tested separately
 - A user can replace connection with connection:new_stream() and experience no problems except maybe performance issues.

Reply:
fixed